cli_tools/gce_vm_image_import/args/args.go

```diff
@@ -0,0 +1,289 @@
+)
+
+// Flags that are validated.
+const (
+	imageFlag          = "image_name"
+	clientFlag         = "client_id"
+	dataDiskFlag       = "data_disk"
+	osFlag             = "os"
+	customWorkflowFlag = "custom_translate_workflow"
+)
+
+// ParsedArguments holds the structured results of parsing CLI arguments.
+type ParsedArguments struct {
+	Img         importer.ImageSpec
+	Env         importer.Environment
+	Translation importer.TranslationSpec
+}
+
+// ParseArgs parses, validates, and populates the CLI arguments that are
+// used by the importer tool.
+func ParseArgs(args []string, populator param.Populator, sf importer.SourceFactory) (ParsedArguments, error) {
+	flagSet := flag.NewFlagSet("image-import", flag.ContinueOnError)
+	// Don't write parse errors to stdout, instead propagate them via an
+	// exception since we use flag.ContinueOnError.
+	flagSet.SetOutput(ioutil.Discard)
+
+	img := importer.ImageSpec{}
+	env := importer.Environment{
+		CurrentExecutablePath: os.Args[0],
+	}
+	translation := importer.TranslationSpec{}
+
+	registerImageSpec(flagSet, &img)
+	registerEnvironment(flagSet, &env)
+	registerTranslationSpec(flagSet, &translation)
+
+	err := flagSet.Parse(args)
+	if err != nil {
+		return ParsedArguments{}, err
+	}
+
+	translation.Source, err = sf.Init(translation.SourceFile, translation.SourceImage)
+	if err != nil {
+		return ParsedArguments{}, err
+	}
+
+	if err := populator.PopulateMissingParameters(&env.Project, &env.Zone, &env.Region,
+		&env.ScratchBucketGcsPath, translation.SourceFile, &img.StorageLocation); err != nil {
+		return ParsedArguments{}, err
+	}
+
+	if err := populateNetwork(&env); err != nil {
+		return ParsedArguments{}, err
+	}
+
+	if err := validate(img, env, translation); err != nil {
+		return ParsedArguments{}, err
+	}
+
+	return ParsedArguments{img, env, translation}, err
+}
+
+func validate(img importer.ImageSpec, env importer.Environment, rules importer.TranslationSpec) error {
+	if env.ClientID == "" {
+		return fmt.Errorf("-%s has to be specified", clientFlag)
+	}
+	if img.Name == "" {
+		return fmt.Errorf("-%s has to be specified", imageFlag)
+	}
+	if !rules.DataDisk && rules.OS == "" && rules.CustomWorkflow == "" {
+		return fmt.Errorf("-%s, -%s, or -%s has to be specified",
+			dataDiskFlag, osFlag, customWorkflowFlag)
+	}
+	if rules.DataDisk && (rules.OS != "" || rules.CustomWorkflow != "") {
+		return fmt.Errorf("when -%s is specified, -%s and -%s should be empty",
+			dataDiskFlag, osFlag, customWorkflowFlag)
+	}
+	if rules.OS != "" && rules.CustomWorkflow != "" {
+		return fmt.Errorf("-%s and -%s can't be both specified",
+			osFlag, customWorkflowFlag)
+	}
+	if rules.OS != "" {
+		if err := daisy_utils.ValidateOS(rules.OS); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func populateNetwork(e *importer.Environment) error {
+	// Populate Network and Subnet. Two goals:
+	//
+	// a. Explicitly use the 'default' network only when
+	//    network is omitted and subnet is empty.
```

Comment:
A blank network has two meanings, depending on whether the subnet is present. When the subnet is present, it means "search for the network associated with this subnet." When the subnet is not present, it means "use default."

So, my goal was to codify our assumptions explicitly, both for future readers, and to identify bugs in those assumptions earlier rather than later.